src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_bo.c

```diff
@@ -1329,3 +1329,70 @@ int xocl_usage_stat_ioctl(struct drm_device *dev, void *data,
 
 	return 0;
 }
+
+static int get_bo_paddr(struct xocl_dev *xdev, struct drm_file *filp,
+	uint32_t bo_hdl, size_t off, size_t size, uint64_t *paddrp)
+{
+	struct drm_device *ddev = filp->minor->dev;
+	struct drm_gem_object *obj;
+	struct drm_xocl_bo *xobj;
+
+	obj = xocl_gem_object_lookup(ddev, filp, bo_hdl);
+	if (!obj) {
+		userpf_err(xdev, "Failed to look up GEM BO 0x%x\n", bo_hdl);
+		return -ENOENT;
+	}
+
+	xobj = to_xocl_bo(obj);
+	if (!xobj->mm_node) {
+		/* Not a local BO */
+		XOCL_DRM_GEM_OBJECT_PUT_UNLOCKED(obj);
+		return -EADDRNOTAVAIL;
+	}
+
+	if (obj->size <= off || obj->size < off + size) {
+		userpf_err(xdev, "Failed to get paddr for BO 0x%x\n", bo_hdl);
+		XOCL_DRM_GEM_OBJECT_PUT_UNLOCKED(obj);
+		return -EINVAL;
+	}
+
+	*paddrp = xobj->mm_node->start + off;
+	XOCL_DRM_GEM_OBJECT_PUT_UNLOCKED(obj);
+	return 0;
+}
+
+int xocl_m2m_bo_ioctl(struct drm_device *dev, void *data,
```

Comment:
if you added back p2p bo copy, this function name should be changed to xocl_copy_bo_ioctl()? it's not just "m2m" any more, right?

Reply:
AI for me, once we discuss the P2P case, I would like to adjust the name to a proper name that we agree.